src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,155 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int SOFT_MAX_ARRAY_LENGTH = Integer.MAX_VALUE - 8;
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final Object[] DEFAULT_ELEMENTS = {};
+    private Object[] elements;
+    private int size;
+
+    public ArrayList() {
+        this.elements = new Object[DEFAULT_CAPACITY];
+    }
+
+    @SuppressWarnings("unchecked")
+    T elements(int index) {
+        return (T) elements[index];
+    }
+
     @Override
     public void add(T value) {
-
+        add(value, size);
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkIndex(index, size);
+        int s;
+        Object[] elementData;
+        if ((s = size) == (elementData = this.elements).length) {
+            elementData = grow(size + 1);
+        }
+        if (index == 0) {
+            System.arraycopy(elementData, 0, elementData, 1, s);
+        } else {
+            System.arraycopy(elementData, 0, elementData, 0, index);
+            System.arraycopy(elementData, index, elementData, index + 1, s - index);
+        }
+        elementData[index] = value;
+        size = s + 1;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (list == null) {
+            return;
+        }
+        for (int i = 0; i < ((ArrayList<T>) list).size; i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index, (size == 0) ? 0 : (size - 1));
+        return elements(index);
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index, (size == 0) ? 0 : (size - 1));
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index, (size == 0) ? 0 : (size - 1));
```

Comment:
```suggestion
        checkIndex(index, (size == 0) ? 0 : (size - 1));
```
you have the same line in several places, let's remove second parameter from this method and pass only index